File: 2025/day7/day7.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

ll dfs2(int row, int col, vector<string>& grid, vector<vector<ll>>& memo, int rows, int cols){
    if (col < 0 || col >= cols) return 0;
    if(row == rows) return 1;
    if(memo[row][col] != -1) return memo[row][col];

    char curr = grid[row][col];
    ll ans;
    if (curr == 'S' || curr == '|' || curr == '.'){
        ans = dfs2(row + 1, col, grid, memo, rows, cols);
    }else if(curr == '^'){
        ans = dfs2(row + 1, col + 1, grid, memo, rows, cols) + dfs2(row + 1, col - 1, grid, memo, rows, cols);
    }else{
        ans = 0;
    }
    return memo[row][col] = ans;
}


int main() {
  ifstream file("day7_inputs.txt");
  vector<string> grid;
  string line;

  while (getline(file, line)) {
    grid.push_back(line);
  }

  int rows = grid.size();
  int cols = grid[0].size();
  vector<vector<ll>> memo;
  memo.assign(rows, vector<ll>(cols, -1));

  int start_col = -1;
  for (int j = 0; j < cols; j++) {
    if (grid[0][j] == 'S') {
      start_col = j;
      break;
    }
  }

  set<pair<int, int>> visited;
  int count = 0;

  // min-heap priority queue: smallest (row, col) comes out first
  priority_queue<pair<int, int>, vector<pair<int, int>>,
                 greater<pair<int, int>>>
      pq;

  pq.push({1, start_col});

  while (!pq.empty()) {
    auto [row, col] = pq.top();
    pq.pop();

    // bounds check
    if (row < 0 || row >= rows || col < 0 || col >= cols)
      continue;

    // visited check
    if (visited.count({row, col}))
      continue;

    visited.insert({row, col});
    char curr = grid[row][col];

    if (curr == '^') {
      count++;
      pq.push({row + 1, col - 1}); // down-left
      pq.push({row + 1, col + 1}); // down-right
    } else if (curr == '.') {
      pq.push({row + 1, col}); // straight down
    }
  }

  cout << "Part ones solution is " << count << '\n';
  ll count1= dfs2(1, start_col, grid, memo, rows, cols);
  cout << "Part two solution is " << count1 << '\n';
  return 0;
}

File: 2025/day7/day7.py
grid = [list(line) for line in open("day7_inputs.txt", "r").read().splitlines()]

rows = len(grid)
cols = len(grid[0])

start_col = grid[0].index("S")

visited = set()
count = 0

# just backtrack for every direction
def backtrack1(row, col):
    global count

    if row < 0 or row >= rows or col < 0 or col >= cols:
        return

    if (row, col) in visited:
        return

    visited.add((row, col))
    curr = grid[row][col]

    if curr == "^":
        count += 1
        backtrack1(row + 1, col - 1)
        backtrack1(row + 1, col + 1)
    elif curr == ".":
        backtrack1(row + 1, col)

backtrack1(1, start_col)
print(f"Part ones solution is {count}")

count = 0
memo = [[-1] * cols for _ in range(rows)]


def backtrack2(row, col):
    global count

    # if we went out side the frame in horizantally
    if col < 0 or col >= cols:
        return 0

    # if the splitter reached the end of the grid
    if row == rows:
        return 1

    if memo[row][col] != -1:
        return memo[row][col]

    ans = 0
    curr = grid[row][col]

    # if there is no splitter just go normally
    if curr == "S"  or curr == "|" or curr == '.':
        ans = backtrack2(row + 1, col)
    elif curr == '^':
        ans = backtrack2(row + 1, col - 1) + backtrack2(row + 1, col + 1)

    memo[row][col] = ans
    return memo[row][col]

count = backtrack2(1, start_col)
print(f"Part2 solution is : {count}")
